### PR TITLE
check for multiple definitions

### PR DESCRIPTION
**Pull request by bdemann**

### CODE REVIEW

File: src/act/abstract_canister_tree.rs [hunk 1]
```diff
@@ -40,11 +43,13 @@ impl AbstractCanisterTree {
             .verify_type_refs_have_corresponding_definitions()
             .err()
             .into_iter()
+            .chain(self.verify_type_refs_defs_are_unique().err())
```
Comment (lastmjs) [inline src/act/abstract_canister_tree.rs:46]: fix

Comment (lastmjs) [reply src/act/abstract_canister_tree.rs:46]: rename to type_defs

File: src/act/abstract_canister_tree.rs [hunk 2]
```diff
@@ -40,11 +43,13 @@ impl AbstractCanisterTree {
...
             .chain(
                 self.verify_guard_function_names_have_corresponding_definitions()
-                    .err()
-                    .into_iter(),
+                    .err(),
             )
+            .chain(self.verify_guard_functions_defs_are_unique().err())
```
Comment (lastmjs) [inline src/act/abstract_canister_tree.rs:51]: guard_function_defs

File: src/act/abstract_canister_tree.rs [hunk 3]
```diff
@@ -264,3 +323,15 @@ impl HasTypeRefs for AbstractCanisterTree {
             .collect()
     }
 }
+
+fn find_duplicates<T: Eq + std::hash::Hash>(list: &[T]) -> Vec<&T> {
```
Comment (lastmjs) [inline src/act/abstract_canister_tree.rs:327]: Immutable:

```rust
fn find_duplicates<T: Eq + Hash>(list: &[T]) -> Vec<&T> {
    let count_map = list.iter().fold(HashMap::new(), |mut acc, item| {
        *acc.entry(item).or_insert(0) += 1;
        acc
    });

    count_map
        .iter()
        .filter(|(_, &count)| count > 1)
        .map(|(&item, _)| item)
        .collect()
}
```

File: src/act/abstract_canister_tree.rs [hunk 4]
```diff
@@ -209,6 +214,60 @@ impl AbstractCanisterTree {
...
+            false => Err(duplicates
+                .into_iter()
+                .map(|type_ref| Error::MultipleTypeDefinitions(type_ref.clone()))
+                .collect()),
+        }
+    }
+
+    fn verify_guard_functions_defs_are_unique(&self) -> Result<(), Vec<Error>> {
+        let defined_names: Vec<_> = self
+            .guard_functions
```
Comment (lastmjs) [inline src/act/abstract_canister_tree.rs:232]: Implement `get_defined_names` on guard_functions, query methods, update methods, and anywhere else it makes sense

File: src/act/abstract_canister_tree.rs
Comment (lastmjs) [inline src/act/abstract_canister_tree.rs:1]: Expand to checking for all other duplicate-possible things

**Review by lastmjs — CHANGES_REQUESTED** [review]

File: src/act/abstract_canister_tree.rs
Comment (bdemann) [reply src/act/abstract_canister_tree.rs:1]: - [x] Canister methods
     - [x] update
     - [x] query
     - [x] ~heartbeat~
     - [x] ~init~
     - [x] ~inspect~
     - [x] ~pre/post upgrade~
- [x] Candid types
- [x] guard functions

Comment (bdemann) [reply src/act/abstract_canister_tree.rs:1]: for heartbeat, init, inspect, and pre/post upgrade, the cdk framework does not have access to the names used to define those. Everything else that does have a name is taken care of.

**Review by lastmjs — APPROVED** [review]

**Pull request merged**